SeQuant/core/optimize.hpp

```diff
@@ -100,31 +100,53 @@ struct OptRes {
   eval_seq_t sequence;
 };
 
-/// returns a pair of index vectors
-/// first element of the pair is the vector of common indices compared by labels
-/// second element of the pair is the set symmetric difference of the input
-/// index vectors if either of the input index container is empty, the result is
-/// a pair of empty vectors
+///
+/// returns a vector of Index objects that are common in @c idxs1 and @c idxs2
+/// that are sorted using Index:LabelCompare{}.
+///
 /// @note I1 and I2 containers are assumed to be sorted by using
 /// Index::LabelCompare{};
 template <typename I1, typename I2>
-std::pair<container::svector<Index>, container::svector<Index>> common_indices(
-    I1 const& idxs1, I2 const& idxs2) {
+container::svector<Index> common_indices(I1 const& idxs1, I2 const& idxs2) {
   using std::back_inserter;
   using std::begin;
   using std::end;
+  using std::set_intersection;
+
+  auto result = container::svector<Index>{};
+
+  set_intersection(begin(idxs1), end(idxs1), begin(idxs2), end(idxs2),
+                   back_inserter(result), Index::LabelCompare{});
+  return result;
+}
 
-  container::vector<Index> i1vec(begin(idxs1), end(idxs1)),
-      i2vec(begin(idxs2), end(idxs2));
-  if (i1vec.empty() || i2vec.empty()) return {{}, {}};
-
-  container::svector<Index> commons, diffs;
-  std::set_intersection(begin(i1vec), end(i1vec), begin(i2vec), end(i2vec),
-                        back_inserter(commons), Index::LabelCompare{});
-  std::set_symmetric_difference(begin(i1vec), end(i1vec), begin(i2vec),
-                                end(i2vec), back_inserter(diffs),
-                                Index::LabelCompare{});
-  return {commons, diffs};
+///
+/// returns a vector of Index objects that are common in @c idxs1 and @c idxs2.
+///
+/// @note I1 and I2 containers are assumed to be sorted by using
+/// Index::LabelCompare{};
+template <typename I1, typename I2>
+container::svector<Index> diff_indices(I1 const& idxs1, I2 const& idxs2) {
+  using std::back_inserter;
+  using std::begin;
+  using std::end;
+  using std::set_symmetric_difference;
+
+  auto result = container::svector<Index>{};
+
+  set_symmetric_difference(begin(idxs1), end(idxs1), begin(idxs2), end(idxs2),
+                           back_inserter(result), Index::LabelCompare{});
+  return result;
+}
+
+///
+/// TODO: Use C++20 <bit> header when possible
+/// \tparam T integral type
+/// \return true if @c x has a single bit on in its bit representation.
+///
+template <typename T>
+bool has_single_bit(T x) noexcept {
+  return x != 0 && (x & (x - 1)) == 0;
```

Comment:
```suggestion
  return std::bitset<sizeof(T)>(x).count() == 1;
```
This is more readable imo (and the compiler should be able to optimize away all overhead this might introduce)

Reply:
There is a function in standard library in C++20. I am trying to emulate that.

https://en.cppreference.com/w/cpp/numeric/has_single_bit

Reply:
I know. But writing bit-fiddling down explicitly usually is reasonably unreadable. 
https://quick-bench.com/q/LSJ_Spc0QvfGRECIMV3RWVwalwQ indicates that there are more readable ways to write this down that only intrdoduce a small overhead in performance. Interestingly, C++20's `has_single_bit` doesn't seem to perform that well. Even a simple `x % 2 == 0` performs better :eyes: 
Anyway, I only wanted to point out that there is a small potential for making the code more readable for non-bit-wizards. Feel free to ignore my comment though, if you don't think anything needs to be changed :)

Reply:
using "standard" names is the way to go. Even when the name is eminently confusing (like `has_single_bit`). Unfortunately we are not ready to _require_ C++20 across the stack (SeQuant is a small part of it) ... not yet at least. But it's a good idea to actually use `std::has_single_bit` directly if C++20 is available (`#if __cplusplus >= ...`)